File: src/exam-parser.ts
import * as cheerio from "cheerio";
import { ExamArrangement, TimetableCourse } from "./types.js";
import { logStep } from "./logger.js";

const normalizeText = (value: string): string =>
  value
    .replace(/\u00a0/g, " ")
    .replace(/&nbsp;/gi, " ")
    .replace(/\r/g, "\n")
    .replace(/[ \t]+/g, " ")
    .replace(/\n{3,}/g, "\n\n")
    .trim();

const cleanInlineText = (value: string): string =>
  normalizeText(value).replace(/\s*\n\s*/g, " ");

const normalizeLookupKey = (value: string): string =>
  cleanInlineText(value).replace(/\s+/g, "").toLowerCase();

const addTeacher = (index: Map<string, Set<string>>, key: string, teacher: string | undefined): void => {
  const normalizedKey = normalizeLookupKey(key);
  const normalizedTeacher = cleanInlineText(teacher ?? "");

  if (!normalizedKey || !normalizedTeacher) {
    return;
  }

  const teachers = index.get(normalizedKey) ?? new Set<string>();
  teachers.add(normalizedTeacher);
  index.set(normalizedKey, teachers);
};

const buildTeacherIndex = (courses: TimetableCourse[]): Map<string, Set<string>> => {
  const index = new Map<string, Set<string>>();

  for (const course of courses) {
    addTeacher(index, course.courseCode ?? "", course.teacher);
    addTeacher(index, course.courseName, course.teacher);
  }

  return index;
};

const findTeacher = (
  teacherIndex: Map<string, Set<string>>,
  courseCode: string,
  courseName: string
): string | undefined => {
  const teachers =
    teacherIndex.get(normalizeLookupKey(courseCode)) ??
    teacherIndex.get(normalizeLookupKey(courseName));

  if (!teachers || teachers.size === 0) {
    return undefined;
  }

  return [...teachers].join("、");
};

export const parseExamArrangementHtml = (
  html: string,
  courses: TimetableCourse[] = []
): ExamArrangement[] => {
  logStep("Parsing exam arrangement HTML.");

  const $ = cheerio.load(html);
  const teacherIndex = buildTeacherIndex(courses);
  const exams: ExamArrangement[] = [];

  $("#dataList tr")
    .toArray()
    .slice(1)
    .forEach((row) => {
      const cells = $(row).children("td").toArray();
      if (cells.length < 7) {
        return;
      }

      const indexText = cleanInlineText($(cells[0]).text());
      const examSession = cleanInlineText($(cells[1]).text());
      const courseCode = cleanInlineText($(cells[2]).text());
      const courseName = cleanInlineText($(cells[3]).text());
      const examTime = cleanInlineText($(cells[4]).text());
      const examRoom = cleanInlineText($(cells[5]).text());
      const seatNumber = cleanInlineText($(cells[6]).text());
      const teacher = findTeacher(teacherIndex, courseCode, courseName);

      if (!courseCode && !courseName && !examTime) {
        return;
      }

      exams.push({
        index: Number(indexText) || exams.length + 1,
        examSession,
        courseCode,
        courseName,
        examTime,
        examRoom,
        seatNumber,
        teacher,
        rawText: [
          indexText,
          examSession,
          courseCode,
          courseName,
          teacher,
          examTime,
          examRoom,
          seatNumber
        ]
          .filter(Boolean)
          .join("\n")
      });
    });

  logStep(`Parsed ${exams.length} exam arrangement entries.`);
  return exams;
};

File: android/app/src/main/java/com/classsche/mobile/ExamParser.kt
package com.classsche.mobile

import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import java.io.ByteArrayInputStream

object ExamParser {
  fun parse(html: String, courses: List<TimetableCourse> = emptyList()): List<ExamArrangement> =
    parseDocument(Jsoup.parse(html), courses)

  fun parseBytes(bytes: ByteArray, baseUri: String, courses: List<TimetableCourse> = emptyList()): List<ExamArrangement> =
    parseDocument(Jsoup.parse(ByteArrayInputStream(bytes), null, baseUri), courses)

  private fun parseDocument(document: Document, courses: List<TimetableCourse>): List<ExamArrangement> {
    val teacherIndex = buildTeacherIndex(courses)
    val exams = mutableListOf<ExamArrangement>()

    val rows = document.select("#dataList tr")
    if (rows.isEmpty()) {
      val title = document.title()
      if (title.contains("登录") || title.contains("login", ignoreCase = true)) {
        throw IllegalStateException("未获取到考试列表，会话可能已过期。当前页面: $title")
      }
      throw IllegalStateException("未在考试页面中找到 #dataList 表格。当前页面: $title")
    }

    rows.drop(1).forEach { row ->
      val cells = row.select("td")
      if (cells.size < 7) return@forEach

      val indexText = cleanInlineText(cells[0].text())
      val examSession = cleanInlineText(cells[1].text())
      val courseCode = cleanInlineText(cells[2].text())
      val courseName = cleanInlineText(cells[3].text())
      val examTime = cleanInlineText(cells[4].text())
      val examRoom = cleanInlineText(cells[5].text())
      val seatNumber = cleanInlineText(cells[6].text())
      val teacher = findTeacher(teacherIndex, courseCode, courseName)

      if (courseCode.isBlank() && courseName.isBlank() && examTime.isBlank()) {
        return@forEach
      }

      exams += ExamArrangement(
        index = indexText.toIntOrNull() ?: (exams.size + 1),
        examSession = examSession,
        courseCode = courseCode,
        courseName = courseName,
        examTime = examTime,
        examRoom = examRoom,
        seatNumber = seatNumber,
        teacher = teacher,
        rawText = listOf(
          indexText,
          examSession,
          courseCode,
          courseName,
          teacher,
          examTime,
          examRoom,
          seatNumber
        ).filter { it.isNotBlank() }.joinToString("\n")
      )
    }

    return exams
  }

  private fun normalizeText(value: String): String =
    value
      .replace('\u00A0', ' ')
      .replace("&nbsp;", " ")
      .replace("\r", "\n")
      .replace(Regex("[ \\t]+"), " ")
      .replace(Regex("\\n{3,}"), "\n\n")
      .trim()

  private fun cleanInlineText(value: String): String =
    normalizeText(value).replace(Regex("\\s*\\n\\s*"), " ")

  private fun normalizeLookupKey(value: String): String =
    cleanInlineText(value).replace(Regex("\\s+"), "").lowercase()

  private fun addTeacher(index: MutableMap<String, MutableSet<String>>, key: String, teacher: String) {
    val normalizedKey = normalizeLookupKey(key)
    val normalizedTeacher = cleanInlineText(teacher)
    if (normalizedKey.isBlank() || normalizedTeacher.isBlank()) {
      return
    }

    index.getOrPut(normalizedKey) { linkedSetOf() }.add(normalizedTeacher)
  }

  private fun buildTeacherIndex(courses: List<TimetableCourse>): Map<String, Set<String>> {
    val index = linkedMapOf<String, MutableSet<String>>()
    courses.forEach { course ->
      addTeacher(index, course.courseCode, course.teacher)
      addTeacher(index, course.courseName, course.teacher)
    }
    return index
  }

  private fun findTeacher(teacherIndex: Map<String, Set<String>>, courseCode: String, courseName: String): String {
    val teachers =
      teacherIndex[normalizeLookupKey(courseCode)]
        ?: teacherIndex[normalizeLookupKey(courseName)]
        ?: return ""

    return teachers.joinToString(" / ")
  }
}
